[Impeller] cache descriptor set layouts.

jonahwilliams opened the pull request:
Currently we reset the entire descriptor pool(s) each frame.

We can actually reuse the allocated descriptor sets, as that is safe to do as soon as the cmd buffer that references them has been submitted (AFAIK). This provides a minor speed up for a large amount of small drawing ops, but not much beyond that.

Commit:
++

Commit:
Merge branch 'master' into cache_pipeline

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.cc:
```diff
@@ -187,15 +171,17 @@ vk::UniqueDescriptorPool DescriptorPoolRecyclerVK::Create() {
+std::shared_ptr<DescriptorPoolVK>
+DescriptorPoolRecyclerVK::GetDescriptorPool() {
+  {
+    Lock recycled_lock(recycled_mutex_);
```
Are we sure this is faster?  We are introducing locks.

jonahwilliams replied:
There were already locks. Before:

1. Raster: Finish work, push to resource manager thread.
2. Resource Manager: call reset, push to context (acquire lock).
3. Raster: Pick recycled pool.

Now we just skip the resource manager part, as we don't reset the pool unless it was completely unused.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/compute_pipeline_vk.h:
```diff
@@ -35,13 +36,18 @@ class ComputePipelineVK final
 
+  /// @brief Retrieve the unique identifier for this pipeline's descriptor set
+  ///        layout.
+  uint64_t GetPipelineKey() const;
```
nit: `using PipelineKey = uint64_t;`

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.h:
```diff
@@ -6,13 +6,21 @@
+  std::vector<vk::DescriptorSet> used;
+};
+
+using DescriptorCacheMap = std::unordered_map<uint64_t, DescriptorCache>;
```
use `absl::flat_hash_map`

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.cc:
```diff
@@ -111,6 +86,10 @@ fml::StatusOr<vk::DescriptorSet> DescriptorPoolVK::AllocateDescriptorSets(
     set_info.setDescriptorPool(pools_.back().get());
     result = context_vk.GetDevice().allocateDescriptorSets(&set_info, &set);
   }
+  if (existing == descriptor_sets_.end()) {
+    descriptor_sets_[pipeline_key] = DescriptorCache{};
+  }
```
use `try_emplace` to avoid double lookups

jonahwilliams replied:
I need some pointers here...

gaaclarke replied:
```c++
descriptor_sets_.try_emplace(pipeline_key, DescriptorCache{});
result.first->used.push_back(set);
```

jonahwilliams replied:
actually `lookup_result.first->second.used.push_back(set);`

Its an iterator to an iterator. What a language.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.cc:
```diff
@@ -130,30 +109,35 @@ fml::Status DescriptorPoolVK::CreateNewPool(const ContextVK& context_vk) {
+  // cached then delete the oldest entry.
+  Lock recycled_lock(recycled_mutex_);
+  if (recycled_.size() >= kMaxRecycledPools) {
+    auto& front = recycled_.front();
+    front->Destroy();
+    recycled_.erase(recycled_.begin());
```
You are erasing from the front which is O(n), can't we remove from the end?

jonahwilliams replied:
Sure, we can just delete the newest entry instead.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.cc:
```diff
@@ -130,30 +109,35 @@ fml::Status DescriptorPoolVK::CreateNewPool(const ContextVK& context_vk) {
+  // Move the pool to the recycled list. If more than 32 pool are
+  // cached then delete the oldest entry.
+  Lock recycled_lock(recycled_mutex_);
+  if (recycled_.size() >= kMaxRecycledPools) {
```
```suggestion
  while (recycled_.size() >= kMaxRecycledPools) {
```

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk_unittests.cc:
```diff
@@ -31,29 +28,10 @@ TEST(DescriptorPoolRecyclerVKTest, ReclaimMakesDescriptorPoolAvailable) {
     // Fetch a pool (which will be created).
     auto pool = DescriptorPoolVK(context);
-    pool.AllocateDescriptorSets({}, *context);
+    pool.AllocateDescriptorSets({}, 0, *context);
```
```suggestion
    pool.AllocateDescriptorSets({}, /*pipeline_key=*/0, *context);
```

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk_unittests.cc:
```diff
@@ -31,29 +28,10 @@ TEST(DescriptorPoolRecyclerVKTest, ReclaimMakesDescriptorPoolAvailable) {
-  }
-
-  auto const pool = context->GetDescriptorPoolRecycler()->Get();
+  auto const pool = context->GetDescriptorPoolRecycler()->GetDescriptorPool();
```
auto const?

jonahwilliams replied:
what about it?

jonahwilliams replied:
oh, const auto?

gaaclarke replied:
> what about it?

I read this in a new york accent lol.  I don't know what `auto const` is.  It seems like it may be functionally equivalent if the deduced type is not a pointer.  Let's just remove the `auto`.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk_unittests.cc:
```diff
@@ -71,50 +49,27 @@ TEST(DescriptorPoolRecyclerVKTest, ReclaimDropsDescriptorPoolIfSizeIsExceeded) {
     for (auto i = 0u; i < 33; i++) {
       auto pool = std::make_unique<DescriptorPoolVK>(context);
-      pool->AllocateDescriptorSets({}, *context);
+      pool->AllocateDescriptorSets({}, 0, *context);
```
```suggestion
      pool->AllocateDescriptorSets({}, /*pipeline_key=*/0, *context);
```

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk_unittests.cc:
```diff
@@ -71,50 +49,27 @@ TEST(DescriptorPoolRecyclerVKTest, ReclaimDropsDescriptorPoolIfSizeIsExceeded) {
-      auto pool = std::make_unique<DescriptorPoolVK>(context);
-      pool->AllocateDescriptorSets({}, *context);
+      auto pool = context->GetDescriptorPoolRecycler()->GetDescriptorPool();
+      pool->AllocateDescriptorSets({}, 0, *context);
```
```suggestion
      pool->AllocateDescriptorSets({}, /*pipeline_key=*/0, *context);
```

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk_unittests.cc:
```diff
@@ -71,50 +49,27 @@ TEST(DescriptorPoolRecyclerVKTest, ReclaimDropsDescriptorPoolIfSizeIsExceeded) {
   // allocated.
   {
-    std::vector<std::unique_ptr<DescriptorPoolVK>> pools;
+    std::vector<std::shared_ptr<DescriptorPoolVK>> pools;
```
Why this change?

jonahwilliams replied:
before we used to vend unique objects that had internal shared state. now we just vend the shared state.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/pipeline_library_vk.cc:
```diff
@@ -179,7 +172,8 @@ PipelineFuture<PipelineDescriptor> PipelineLibraryVK::GetPipeline(
   auto weak_this = weak_from_this();
 
-  auto generation_task = [descriptor, weak_this, promise]() {
+  uint64_t next_key = pipeline_key_++;
```
Is `GetPipeline` guaranteed to be called from the same thread?

jonahwilliams replied:
I added a guard that verifies that is synchronized by the same mutex that guards the pipeline map.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/pipeline_library_vk.h:
```diff
@@ -46,6 +46,7 @@ class PipelineLibraryVK final
   ComputePipelineMap compute_pipelines_ IPLR_GUARDED_BY(
       compute_pipelines_mutex_);
   std::atomic_size_t frames_acquired_ = 0u;
+  uint64_t pipeline_key_ = 0;
```
I'd start counting at 1 instead of at zero to keep uninitialized data to mean it's invalid.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/pipeline_vk.h:
```diff
@@ -62,6 +65,7 @@ class PipelineVK final
   vk::UniquePipelineLayout layout_;
   vk::UniqueDescriptorSetLayout descriptor_set_layout_;
   std::shared_ptr<SamplerVK> immutable_sampler_;
+  uint64_t pipeline_key_;
```
```suggestion
  const uint64_t pipeline_key_;
```

jonahwilliams replied:
Done

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/descriptor_pool_vk.h:
There are no new tests being added for this functionality.

jonahwilliams replied:
I added a test that shows the descriptors being recycled. Otherwise, everything not exploding should count for something I hope.

gaaclarke reviewed (commented):
Mostly looks good, here's my high level feedback:
1) This is adding a fair bit of complexity around code that relies on thread safety, have you measured the difference so we can justify the risk and added complexity?
1) There are a few places where I wonder if they are thread-safe or not
1) There is new functionality being added that could have test coverage.

gaaclarke commented on engine/src/flutter/impeller/renderer/backend/vulkan/context_vk.cc:
```diff
@@ -546,9 +546,8 @@ std::shared_ptr<CommandBuffer> ContextVK::CreateCommandBuffer() const {
-      descriptor_pool =
-          (cached_descriptor_pool_[std::this_thread::get_id()] =
-               std::make_shared<DescriptorPoolVK>(weak_from_this()));
+      descriptor_pool = (cached_descriptor_pool_[std::this_thread::get_id()] =
+                             descriptor_pool_recycler_->GetDescriptorPool());
```
Here we are doing a lookup based on what thread we are on.  Can we expand that to include the recycler too so we don't have to introduce mutexes?

jonahwilliams replied:
This change doesn't introduce any new mutexes

Commit:
aaron review.

jonahwilliams commented:
Haven't addressed all feedback yet. RE: performance, yes - but let me pull some numbers

Recycling descriptor sets is in the general arm/adreno guidance if you want that too

jonahwilliams commented:
oops, push rejected ... sec...

Commit:
++

Commit:
Merge branch 'cache_pipeline' of github.com:jonahwilliams/flutter into cache_pipeline

jonahwilliams commented:
It looks like using absl::flat_hash_map regresses performance. maybe it has a bad default hash for a sequentially incremented int? So I switched back to unordered map.

Anyway. Running the qr.flutter example app:


Before:

![image](https://github.com/user-attachments/assets/3a1535aa-71a6-4b4e-a6c4-3c8fc616725f)


After:

![image](https://github.com/user-attachments/assets/a7f96ce7-4a91-4fec-9eb1-89ca3c03ff66)

Commit:
back to unordered_map.

Commit:
++

gaaclarke commented:
> It looks like using absl::flat_hash_map regresses performance. maybe it has a bad default hash for a sequentially incremented int? So I switched back to unordered map.

That's surprising.  Absl's recommendation is flat_hash_map by default and the [benchmarks](https://martin.ankerl.com/2022/08/27/hashmap-bench-01/) show it being faster for every usage.

<img width="1750" alt="Screenshot 2025-03-13 at 11 44 04 AM" src="https://github.com/user-attachments/assets/873bdc65-4888-403d-93df-6edddcad0ff2" />

gaaclarke commented:
Your profile that has flat_hash_map, may just be variance in your measurement.  I'm pretty sure flat_hash_map is faster and more memory efficient.

gaaclarke reviewed (approved):
LGTM modulo the `try_emplace`, usage of `auto`, and I recommend the flat_hash_map since that is faster, but the difference is so little it isn't showing up past the standard test variance.  So, not a huge thing, up to you.

jonahwilliams commented:
The difference is pretty consistently faster for me. might be usage based.

Commit:
aaron review.

jmagman commented:
Reason for revert: this caused a video stream flicker regression https://github.com/flutter/flutter/issues/168114
